Fix wrong slicing of a list of 1 dim ndarrays

louvervecken opened the pull request:
`img = img_line[:][0:cols]` was not slicing to the number of
columns but was slicing the rows in stead. It's now re-written
to use a 2-dim numpy array iso a python list of 1-dim numpy arrays.
This 2-dim numpy array uses the same base array the `img` to save
memory and it is sliced to the correct size before being passed to
code that does the actual FFT calculation.

fariza commented:
@louvervecken thanks for the contribution, please add a test that fails with the current implementation and passes with this new one

louvervecken commented:
@fariza I've added a test which fails for the old implementation. 
I also changed the 'masked copy' to a regular copy to prevent non masked ndarrays to change into masked instances.

Something I havent taken into account is one dimensional images, but it seems like this is not working for the FFT function anyway because the fft is calculated on axis 1?...

fariza commented:
@bebuch can you review this?

fariza commented:
> Something I havent taken into account is one dimensional images, but it seems like this is not working for the FFT function anyway because the fft is calculated on axis 1?...

For line cameras, the standard request to create a fake 2d array by adding multiple lines one after the other, so I don't think the FFT is needed for 1d arrays

bebuch commented:
I would propose to outsource the masked array stuff into a separate function. (Name: `row_wise_compressed`?)

The `GetImgShape` should be done after the `if`.

```python

def row_wise_compressed(img):
    img = img.copy()
    # use the same ndarray to save memory
    compressed = np.ma.getdata(img)
    # the maximum numbers or cols is the shape
    cols = compressed.shape[1]
    # keep count of non empty lines
    lines = 0
    # compress line by line
    for line in img:
        # drop masked data
        line = line.compressed()
        # assure the line is not empty
        if line.size == 0:
            continue
        # set the data in the compressed array
        compressed[lines, :line.size] = line
        # keep count of non-empty lines
        lines += 1
        # keep trac of minimum column size
        cols = min(cols, line.size)
    # trim to actual number of lines and columns
    return compressed[:lines, :cols]

# […]

# remove masked entries if img is a masked array
if np.ma.isMaskedArray(img):
    img = row_wise_compressed(img)

_rows, cols = GetImgShape(img)
```

louvervecken commented:
Hi @bebuch, I've made the changes you proposed. I also renamed the argument `img_` to `img`, the trailing underscore is used to protect python keywords but `img` isn't any as far as i know?

bebuch commented:
Hi @louvervecken, looks good for me now :+1: 

I double checked that the parameter name change does not conflict with named parameters when called elsewhere in the project. This is definitely better now, especially since it didn't match the doc before your change.

Btw., sorry for the delay, was on vacation for two days at short notice at the end of last week. ;-)

kirchhoefer1288 reviewed (approved):
Looks also good for me

bebuch commented:
@fariza is currently the only one with merge rights. It may be that he is currently on vacation. Not quite sure what he said at the last meeting.

I assume that no further commits will be added to this pull request. Therefore, you can create a new branch from your current branch and continue working on that and make a new pull request. Since both branches then contain the commits from this pull request, Git should be able to cleanly resolve that automatically once this pull request got merged.
